### PR TITLE
Multi diffusion guidance

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/diffusion/multi_diffusion/dps_guidance.py [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+                retain_graph=self.retain_graph,
+                create_graph=self.create_graph,
+            )
+
+        grad_x = grads[0]
+        if grad_x is None:
+            raise RuntimeError(
+                "Gradient is None. Ensure x has requires_grad=True and that "
+                "x_0 was computed from x."
+            )
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/diffusion/multi_diffusion/dps_guidance.py]: Good catch, the guard was indeed unreachable. Dropped both blocks in bbe58840 rather than passing `allow_unused=True`, the default behavior (PyTorch raising directly) is fine here.

File: physicsnemo/diffusion/multi_diffusion/dps_guidance.py [hunk 2]
```diff
@@ -0,0 +1,870 @@
...
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/diffusion/multi_diffusion/dps_guidance.py]: Same as above, dropped this guard too in bbe58840.

**Review by pzharrington — APPROVED** [review]
Approving to get into rc, will add comments later

File: physicsnemo/diffusion/guidance/dps_guidance.py [hunk 3]
```diff
@@ -25,6 +25,16 @@
 from physicsnemo.diffusion.base import Predictor
 
 
+def _lp_loss_fn(p: int) -> Callable[[Tensor, Tensor], Tensor]:
```
Comment (pzharrington) [inline physicsnemo/diffusion/guidance/dps_guidance.py:28]: Add jaxtpying annotations to these `Tensor`s if possible? (and where it's used below)

File: physicsnemo/diffusion/multi_diffusion/dps_guidance.py [hunk 4]
```diff
@@ -0,0 +1,1015 @@
...
+        alpha_t = self.alpha_fn(t_bc)
+        variance = self.std_y**2 + self.gamma * (sigma_t**2) / (alpha_t**2)
+
+        g = -grad_x / (2 * variance)
+        if slice_start is None and self.fuse:
+            return self.predictor.fuse_fn(g)
+        return g
+
+
+class MultiDiffusionDataConsistencyDPSGuidance:
```
Comment (CharlelieLrt) [inline physicsnemo/diffusion/multi_diffusion/dps_guidance.py:710]: Note to myself: should inherit the protocol for more clarity. Should also apply to the `MultiDiffusionmodelConsistencyDPSGuidance`

File: physicsnemo/diffusion/multi_diffusion/models.py [hunk 5]
```diff
@@ -401,6 +405,11 @@ def __init__(
         self._patching: RandomPatching2D | GridPatching2D | None = None
         self._fuse: bool = False
         self._skip_positional_embedding_injection: bool = False
+        # Persistent buffer so that patch_shape survives checkpoint save/load.
+        # Zeros sentinel means "not yet configured".
+        self.register_buffer(
+            "_patch_shape", torch.zeros(2, dtype=torch.long), persistent=True
```
Comment (pzharrington) [inline physicsnemo/diffusion/multi_diffusion/models.py:411]: Why would we want to embed the patch shape in a checkpoint?

Comment (CharlelieLrt) [reply physicsnemo/diffusion/multi_diffusion/models.py:411]: That's a good question. I spend some time thinking about this and the main reason is because, for example if a SongUNet was trained with a given patch shape, it is unlikely to give good results on a different patch shape.

The current workflow for inference is to call `set_patching()`, and this requires a patch shape. There are two options:
1. Leave the responsability to the user to know whoich patch shape was used during training and call `set_patching(patch_shape=patch_shape)`
2. Makes it slightly simpler for the user, and save the patch shape in the checkpoint.

The current mechanism in the `MultiDiffusionPredicor` supports both, as it allows to change the patch shape for inference (this will raise a warning if it differs from the patch shape saved in the checkpoint used during training, but it is possible). If nothing is provided it defaults to the patch shape saved in the checkpoint.

File: physicsnemo/diffusion/multi_diffusion/predictor.py [hunk 6]
```diff
@@ -153,83 +205,532 @@ class MultiDiffusionPredictor(Predictor):
...
+        self._md_model._skip_positional_embedding_injection = True
+        # Internal model fusing is handled externally by this predictor (via
+        # fuse_fn); keep it disabled so gradient checkpointing replays observe
+        # a stable model state across forward and backward passes.
+        self._md_model._fuse = False
+
+        # Prediction-type conversion (same pattern as MultiDiffusionMSEDSMLoss).
+        match prediction_type:
+            case "x0":
+                self._to_x0: Callable[[Tensor, Tensor, Tensor], Tensor] = (
```
Comment (pzharrington) [inline physicsnemo/diffusion/multi_diffusion/predictor.py:346]: jaxtyping?

File: physicsnemo/diffusion/multi_diffusion/predictor.py [hunk 7]
```diff
@@ -153,83 +205,532 @@ class MultiDiffusionPredictor(Predictor):
...
+                self._to_x0 = epsilon_to_x0_fn
+            case _:
+                raise ValueError(
+                    f"prediction_type must be 'x0', 'score', or 'epsilon', "
+                    f"got '{prediction_type}'."
+                )
+
+        # Bind the model-call helper once so the use_checkpointing branch is
+        # resolved at construction rather than on every forward call.
+        self._call_model: Callable[[Tensor, Tensor, Any], Tensor] = (
```
Comment (pzharrington) [inline physicsnemo/diffusion/multi_diffusion/predictor.py:369]: jaxtyping?

File: physicsnemo/diffusion/multi_diffusion/models.py (hunk 5 above)
Comment (pzharrington) [reply physicsnemo/diffusion/multi_diffusion/models.py:411]: Ok, yeah I think this makes sense
